lib/rspec-puppet/adapters.rb

```diff
@@ -194,6 +202,24 @@ def settings_map
     end
 
     class Adapter6X < Adapter40
+      def setup_puppet(example_group)
+        case RSpec.configuration.facter_implementation.to_sym
+        when :rspec
+          if Puppet.respond_to?(:runtime) && Puppet.runtime.instance_variable_get(:@runtime_services).key?(:facter)
```

Comment:
Looks like this is implemented via `Puppet::Runtime` rather than a real hash. Perhaps it's safer to see if the key returns anything at all?

```suggestion
          if Puppet.respond_to?(:runtime) && !Puppet.runtime[:facter].nil?
```

I'd also prefer less indented code, so:
```ruby
if !Puppet.respond_to?(:runtime) || Puppet.runtime[:facter].nil?
  raise "Runtime implementations are not supported in Puppet #{Puppet.version}"
end
Puppet.runtime...
```

Reply:
Unfortunately due to the way `Puppet::Runtime` is implemented we can't do what you suggest, because accessing an absent service will raise https://github.com/puppetlabs/puppet/blob/main/lib/puppet/runtime.rb#L41 ... and I think catching exceptions will be even more slower.

Reply:
How about something like:
```ruby
def supports_facter_runtime?
  unless defined?(@supports_facter_runtime)
    begin
      Puppet.runtime[:facter]
      @supports_facter_runtime = true
    rescue ...
      @supports_facter_runtime = false
    end
  end

  @supports_facter_runtime
end

Reply:
Nice! Memoizing will definitely speed things up.